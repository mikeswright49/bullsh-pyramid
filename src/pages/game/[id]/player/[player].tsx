import React from 'react';
import { useRouter } from 'next/router';
import { useGameState } from 'src/hooks/use-game-state';
import { Layout } from 'src/components/layout/layout';
import { GameBoard } from 'src/components/game-board/game-board';
import { usePlayer } from 'src/hooks/use-player';
import { GameStage } from 'src/enums/game-stage';
import { Card } from 'src/components/card/card';

export default function Player(): JSX.Element {
    const router = useRouter();
    const gameId: string = router.query.id as string;
    const playerId: string = router.query.player as string;

    const gameState = useGameState(gameId);
    const player = usePlayer(playerId);

    const { gameStage, activeRow, activeIndex } = gameState;

    function PlayerDisplay() {
        switch (gameStage) {
            case GameStage.Initiation:
                return (
                    <>
                        <div className="stack-y-2">
                            <h2>Waiting for the game to start</h2>
                        </div>
                    </>
                );
            case GameStage.Memorization:
                return (
                    <>
                        <div className="stack-y-2">
                            <h2>You have 30s to remember your cards!</h2>
                        </div>
                        <GameBoard gameState={gameState} />
                    </>
                );
            case GameStage.Flipping:
                return (
                    <>
                        <div className="stack-y-2">
                            <h2>Flipping a card</h2>
                        </div>
                        <GameBoard gameState={gameState} />
                    </>
                );
            case GameStage.Declaration:
                return (
                    <>
                        <div className="stack-y-2">
                            <h2>Quick 5s! Do you have this card?</h2>
                            <Card card={gameState.tiers[activeRow][activeIndex]} />
                        </div>
                        <GameBoard gameState={gameState} />
                    </>
                );
            case GameStage.Bullshit:
                return (
                    <>
                        <div className="stack-y-2">
                            <h2>Did anyone have this card?</h2>
                            <Card card={gameState.tiers[activeRow][activeIndex]} />
                        </div>
                        <GameBoard gameState={gameState} />
                    </>
                );
            case GameStage.Memory:
                return (
                    <>
                        <div className="stack-y-2">
                            <h2>Well which cards did you have?</h2>
                        </div>
                    </>
                );
            default:
                return <h1>Terrible error</h1>;
        }
    }

    return (
        <Layout>
            <PlayerDisplay />
        </Layout>
    );
}
