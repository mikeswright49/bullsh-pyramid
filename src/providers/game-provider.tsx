import React, { useEffect, useState } from 'react';
import { GameState } from 'types/game-state';
import { GameStore } from 'src/stores/game-store';
import { GameStage } from 'src/enums/game-stage';

export const DEFAULT_GAME_STATE: GameState = {
    id: '',
    gameStage: GameStage.Initiation,
    activeRow: 0,
    activeIndex: 0,
    cardCount: 4,
    flipDelay: 5000,
    players: [],
    tiers: [],
};
export const GameContext = React.createContext<GameState>(DEFAULT_GAME_STATE);

export function GameProvider(props: React.PropsWithChildren<{ gameId: string }>) {
    const { gameId } = props;
    const [gameState, setGameState] = useState<GameState>(DEFAULT_GAME_STATE);

    useEffect(() => {
        if (!gameId) {
            return;
        }

        GameStore.subscribeToGame(gameId, (updatedGameState) => {
            setGameState(updatedGameState);
        });

        return () => GameStore.unsubscribeToGame();
    }, [gameId]);

    return <GameContext.Provider value={gameState}>{props.children}</GameContext.Provider>;
}
