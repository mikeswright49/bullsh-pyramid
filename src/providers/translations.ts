export const TRANSLATIONS = {
    en: {
        'game.board.current.row': 'Current Row: {{ activeRow }}',
        'game.board.current.card': 'Current card',
        'game.display.stage.memorization.text': 'You have 30s to remember your cards!',
        'game.display.stage.flipping.text': 'The host is flipping a card',
        'game.display.stage.declaration.text': 'Let the world know if you have it',
        'game.display.stage.assign.text': 'Players are beginning to lie left right and center',
        'game.display.stage.bullshit.text': 'People should know who you think was lieing',
        'game.display.stage.reveal.text': 'We all now know the truth',
        'game.display.stage.memory.text': 'Well which cards did you have?',
        'game.display.stage.complete.text': 'Well time for a new game',
    },
};
