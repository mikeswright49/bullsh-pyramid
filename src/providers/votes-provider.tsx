import React from 'react';
import { useState, useEffect } from 'react';
import { Vote } from 'types/vote';
import { VoteStore } from 'src/stores/vote-store';
export const VotesContext = React.createContext<Vote[]>([]);

export function VotesProvider(props: React.PropsWithChildren<{ gameId: string }>) {
    const [votes, setVotes] = useState([]);
    useEffect(() => {
        if (!votes) {
            return;
        }

        VoteStore.subscribeToVotes(props.gameId, (vote: Vote) => {
            votes[vote.id] = vote;
            setVotes({ ...votes });
        });
    }, [props.gameId]);

    return (
        <VotesContext.Provider value={Object.values(votes)}>{props.children}</VotesContext.Provider>
    );
}
