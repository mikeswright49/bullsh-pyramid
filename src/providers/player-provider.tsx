import React from 'react';
import { useState, useEffect } from 'react';
import { Player } from 'types/player';
import { PlayerStore } from 'src/stores/player-store';
export const PlayerContext = React.createContext<Player>(null);

export function PlayerProvider(props: React.PropsWithChildren<{ playerId: string }>) {
    const [player, setPlayer] = useState<Player>();

    useEffect(() => {
        if (!props.playerId) {
            return;
        }
        PlayerStore.subscribeToPlayer(props.playerId, (updatedPlayer) => {
            setPlayer(updatedPlayer);
        });

        return () => PlayerStore.unsubscribeToPlayer();
    }, [props.playerId]);

    return <PlayerContext.Provider value={player}>{props.children}</PlayerContext.Provider>;
}
