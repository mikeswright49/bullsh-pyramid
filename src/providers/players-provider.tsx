import React from 'react';
import { useState, useEffect } from 'react';
import { Player } from 'types/player';
import { GameStore } from 'src/stores/game-store';

export const PlayersContext = React.createContext<Player[]>([]);

export function PlayersProvider(props: React.PropsWithChildren<{ gameId: string }>) {
    const [players, setPlayers] = useState({});
    useEffect(() => {
        GameStore.subscribeToPlayers(props.gameId, (player: Player) => {
            players[player.id] = player;
            setPlayers({ ...players });
        });
    }, [props.gameId]);
    return (
        <PlayersContext.Provider value={Object.values(players)}>
            {props.children}
        </PlayersContext.Provider>
    );
}
