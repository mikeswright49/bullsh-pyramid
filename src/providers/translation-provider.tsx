import React, { useContext } from 'react';
import { useEffect, useState } from 'react';
import { GameContext } from './game-provider';
import { TRANSLATIONS } from '../translations/translations';

const INTERPOLATE_REGEX = /{{.*?}}/gi;
const INTERPOLATE_PROPERTY_REGEX = /{{(.*)?}}/i;
const INTERPOLATE_CAPTURE_COUNT = 2;

export type Translations = {
    translations: { [key: string]: string };
    translate: (key: string, variables?: { [key: string]: string | number | boolean }) => string;
};

export const TranslationContext = React.createContext<Translations>({
    translate: (a) => a,
    translations: {},
});

export function TranslationProvider(props: React.PropsWithChildren<unknown>) {
    const { language } = useContext(GameContext);
    const [translations, setTranslations] = useState({});
    useEffect(() => {
        setTranslations(TRANSLATIONS[language || 'en']);
    }, [language]);

    /**
     * Returns the nested property value (userMetadata.giftcard.bacon)
     * @param source property value
     */
    function getProperty(source: string[], parent: unknown): string {
        if (!source || !parent) {
            return;
        }
        if (source.length > 1) {
            const property: string = source.shift();
            return this.getProperty(source, parent[property.trim()]);
        }
        return parent[source[0].trim()];
    }

    /**
     * Evaluates the {{}} interpolated strings
     * @param source Source content to be parsed
     * @param data Data object to read values off of
     */
    function evaluateInterpolate(
        source: string,
        data: { [key: string]: string | number | boolean }
    ): string {
        if (!data) {
            return source;
        }

        const matches = source.match(INTERPOLATE_REGEX) || [];
        matches.forEach((match: string) => {
            const properties = INTERPOLATE_PROPERTY_REGEX.exec(match);
            if (properties && properties.length === INTERPOLATE_CAPTURE_COUNT) {
                const property = getProperty(properties[1].split('.'), data);
                source = source.replace(match, property || '');
            }
        });
        return source;
    }

    function translate(key: string, variables?: { [key: string]: string | number | boolean }) {
        if (!translations) {
            return '';
        }
        return evaluateInterpolate(translations[key], variables);
    }

    return (
        <TranslationContext.Provider value={{ translations, translate }}>
            {props.children}
        </TranslationContext.Provider>
    );
}
