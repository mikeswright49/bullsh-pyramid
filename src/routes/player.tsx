import React, { useContext } from 'react';
import { PlayerContext } from 'src/providers/player-provider';
import { HostDisplay } from 'src/components/host-display/host-display';
import { PlayerDisplay } from 'src/components/player-display/player-display';
import { GameBoard } from 'src/components/game-board/game-board';

export function PlayerRoute() {
    const player = useContext(PlayerContext);
    if (!player) {
        return null;
    }
    return (
        <div className="container" data-testid="player-route">
            <div className="row">
                <div className="col-3">
                    {player.isHost && <HostDisplay />}
                    <PlayerDisplay player={player} />
                </div>
                <div className="col-9">
                    <GameBoard />
                </div>
            </div>
        </div>
    );
}
