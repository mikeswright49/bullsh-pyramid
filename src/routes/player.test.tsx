import React from 'react';
import { PlayerRoute } from './player';
import { render } from '@testing-library/react';
import { PlayerContext } from 'src/providers/player-provider';
import { MOCK_PLAYER } from 'src/__tests__/data/mock-player';

describe('Player Route', () => {
    it('should be able to render the component', () => {
        const { container } = render(
            <PlayerContext.Provider value={MOCK_PLAYER}>
                <PlayerRoute />
            </PlayerContext.Provider>
        );
        expect(container.querySelector('[data-testid="player-route"]')).toMatchSnapshot();
    });
});
