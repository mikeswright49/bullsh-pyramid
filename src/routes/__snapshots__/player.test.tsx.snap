// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player Route should be able to render the component 1`] = `
<div
  class="container"
  data-testid="player-route"
>
  <div
    class="row"
  >
    <div
      class="col-3"
    >
      <div
        class="stack-y-2"
      >
        <h2>
          You're the host
        </h2>
        <button
          class="pure-button pure-button-primary"
        >
          Start game
        </button>
      </div>
      <div
        class="stack-y-2"
      >
        <h3>
          Waiting for the game to start
        </h3>
      </div>
    </div>
    <div
      class="col-9"
    >
      <span
        data-testid="at-game-board"
      />
    </div>
  </div>
</div>
`;
